File: repertoire/concepts/data-integration/providers/quality_rule/Freshness.swift
// Quality Rule Provider: Freshness Validation
// Ensures data timestamps are within an acceptable recency window.
// Dimension: timeliness

import Foundation

public final class FreshnessQualityProvider {
    public init() {}

    public func validate(value: Any?, field: FieldDef, record: [String: Any], config: RuleConfig) -> RuleResult {
        let timestampField = config.options?["timestampField"] as? String ?? field.name
        let rawTimestamp: Any?

        if timestampField == field.name {
            rawTimestamp = value
        } else {
            rawTimestamp = record[timestampField]
        }

        guard let rawTs = rawTimestamp, !(rawTs is NSNull) else {
            return RuleResult(
                valid: false,
                message: "Freshness check for '\(field.name)': timestamp field '\(timestampField)' is missing.",
                severity: "warning"
            )
        }

        guard let timestamp = parseTimestamp(rawTs) else {
            return RuleResult(
                valid: false,
                message: "Freshness check for '\(field.name)': cannot parse timestamp value.",
                severity: "error"
            )
        }

        guard let maxAgeMs = parseMaxAge(config.options?["maxAge"]) else {
            return RuleResult(
                valid: false,
                message: "Freshness rule for '\(field.name)' is misconfigured: invalid or missing maxAge.",
                severity: "warning"
            )
        }

        let now = Date()
        let ageMs = now.timeIntervalSince(timestamp) * 1000.0

        if ageMs > maxAgeMs {
            let ageHours = round(ageMs / 3_600_000.0 * 10) / 10
            let maxAgeHours = round(maxAgeMs / 3_600_000.0 * 10) / 10
            return RuleResult(
                valid: false,
                message: "Field '\(field.name)' data is stale: age is \(ageHours)h, maximum allowed is \(maxAgeHours)h.",
                severity: "error"
            )
        }

        return RuleResult(valid: true, message: nil, severity: "info")
    }

    private func parseTimestamp(_ value: Any) -> Date? {
        if let date = value as? Date {
            return date
        }
        if let n = value as? Double {
            // Assume seconds if small, milliseconds if large
            let interval = n > 1_000_000_000_000 ? n / 1000.0 : n
            return Date(timeIntervalSince1970: interval)
        }
        if let n = value as? Int {
            let interval = n > 1_000_000_000_000 ? Double(n) / 1000.0 : Double(n)
            return Date(timeIntervalSince1970: interval)
        }
        if let s = value as? String {
            let isoFormatter = ISO8601DateFormatter()
            isoFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = isoFormatter.date(from: s) { return date }
            let simpleFormatter = DateFormatter()
            simpleFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
            if let date = simpleFormatter.date(from: s) { return date }
            simpleFormatter.dateFormat = "yyyy-MM-dd"
            return simpleFormatter.date(from: s)
        }
        return nil
    }

    private func parseMaxAge(_ maxAge: Any?) -> Double? {
        guard let maxAge = maxAge else { return nil }

        if let n = maxAge as? Int {
            return Double(n) * 1000.0
        }
        if let n = maxAge as? Double {
            return n * 1000.0
        }
        if let s = maxAge as? String {
            return parseDurationString(s)
        }
        return nil
    }

    private func parseDurationString(_ s: String) -> Double? {
        let trimmed = s.trimmingCharacters(in: .whitespaces)

        // Split into numeric and unit parts
        var numStr = ""
        var unitStr = ""
        var foundAlpha = false
        for c in trimmed {
            if c.isLetter {
                foundAlpha = true
                unitStr.append(c)
            } else if !foundAlpha {
                numStr.append(c)
            }
        }

        guard let amount = Double(numStr.trimmingCharacters(in: .whitespaces)) else { return nil }
        let unit = unitStr.trimmingCharacters(in: .whitespaces).lowercased()

        if unit.hasPrefix("s") { return amount * 1000.0 }
        if unit.hasPrefix("mi") || unit == "m" { return amount * 60_000.0 }
        if unit.hasPrefix("h") { return amount * 3_600_000.0 }
        if unit.hasPrefix("d") { return amount * 86_400_000.0 }

        return nil
    }

    public func appliesTo(field: FieldDef) -> Bool {
        let dateTypes = ["date", "datetime", "timestamp"]
        return dateTypes.contains(field.type_.lowercased())
    }

    public func dimension() -> QualityDimension {
        return .timeliness
    }
}

File: repertoire/concepts/data-integration/providers/quality_rule/freshness.ts
// Quality Rule Provider: Freshness Validation
// Ensures data timestamps are within an acceptable recency window.
// Dimension: timeliness

export const PROVIDER_ID = 'freshness';
export const PLUGIN_TYPE = 'quality_rule';

export interface FieldDef {
  name: string;
  type: string;
  required?: boolean;
  constraints?: Record<string, unknown>;
}

export interface RuleConfig {
  options?: Record<string, unknown>;
  threshold?: number;
}

export interface RuleResult {
  valid: boolean;
  message?: string;
  severity: 'error' | 'warning' | 'info';
}

export type QualityDimension = 'completeness' | 'uniqueness' | 'validity' | 'consistency' | 'timeliness' | 'accuracy';

export class FreshnessQualityProvider {
  validate(
    value: unknown,
    field: FieldDef,
    record: Record<string, unknown>,
    config: RuleConfig
  ): RuleResult {
    const timestampField = (config.options?.timestampField as string) ?? field.name;
    const rawTimestamp = timestampField === field.name ? value : record[timestampField];

    if (rawTimestamp === null || rawTimestamp === undefined) {
      return {
        valid: false,
        message: `Freshness check for '${field.name}': timestamp field '${timestampField}' is missing.`,
        severity: 'warning',
      };
    }

    const timestamp = this.parseTimestamp(rawTimestamp);
    if (timestamp === null) {
      return {
        valid: false,
        message: `Freshness check for '${field.name}': cannot parse timestamp value '${rawTimestamp}'.`,
        severity: 'error',
      };
    }

    const maxAge = this.parseMaxAge(config.options?.maxAge);
    if (maxAge === null) {
      return {
        valid: false,
        message: `Freshness rule for '${field.name}' is misconfigured: invalid or missing maxAge.`,
        severity: 'warning',
      };
    }

    const now = Date.now();
    const ageMs = now - timestamp;

    if (ageMs > maxAge) {
      const ageHours = Math.round(ageMs / 3600000 * 10) / 10;
      const maxAgeHours = Math.round(maxAge / 3600000 * 10) / 10;
      return {
        valid: false,
        message: `Field '${field.name}' data is stale: age is ${ageHours}h, maximum allowed is ${maxAgeHours}h.`,
        severity: 'error',
      };
    }

    return { valid: true, severity: 'info' };
  }

  private parseTimestamp(value: unknown): number | null {
    if (typeof value === 'number') return value;
    if (value instanceof Date) {
      const ts = value.getTime();
      return isNaN(ts) ? null : ts;
    }
    if (typeof value === 'string') {
      const parsed = Date.parse(value);
      return isNaN(parsed) ? null : parsed;
    }
    return null;
  }

  private parseMaxAge(maxAge: unknown): number | null {
    if (typeof maxAge === 'number') return maxAge * 1000;

    if (typeof maxAge === 'string') {
      const match = maxAge.match(/^(\d+(?:\.\d+)?)\s*(s|sec|seconds?|m|min|minutes?|h|hours?|d|days?)$/i);
      if (!match) return null;
      const amount = parseFloat(match[1]);
      const unit = match[2].toLowerCase();

      if (unit.startsWith('s')) return amount * 1000;
      if (unit.startsWith('mi') || unit === 'm') return amount * 60 * 1000;
      if (unit.startsWith('h')) return amount * 3600 * 1000;
      if (unit.startsWith('d')) return amount * 86400 * 1000;
    }

    return null;
  }

  appliesTo(field: FieldDef): boolean {
    const dateTypes = ['date', 'datetime', 'timestamp'];
    return dateTypes.includes(field.type.toLowerCase());
  }

  dimension(): QualityDimension {
    return 'timeliness';
  }
}

export default FreshnessQualityProvider;
